pages/api/users/index.ts

```diff
@@ -1,13 +1,86 @@
 import { NextApiRequest, NextApiResponse } from "next";
-import { sampleUserData } from "utils/sample-data";
+import { PrismaClient } from "@prisma/client";
+import sanitizeUser from "utils/sanitizeUser";
+import hash from "utils/hashPassword";
+import { SanitizedUser } from "interfaces";
+import Joi from "joi";
+import { getInviteById } from "../invites/[id]";
 
-const handler = (_: NextApiRequest, res: NextApiResponse): void => {
+const prisma = new PrismaClient();
+
+/**
+ * All users signing up
+ */
+type CreateUserDTO = {
+  name: string;
+  email: string;
+  password: string;
+  inviteCodeId?: string;
+};
+
+/**
+ * Update existing user record with given password by fetching from user invite table (if signing up with an invite code)
+ * OR create an account with given email and password (if user is signing up without invite code)
+ * @param user - contains email, password, and an optional invite code
+ */
+
+export const createAccount = async (
+  user: CreateUserDTO
+): Promise<SanitizedUser | null> => {
+  const loginInfo = {
+    name: user.name,
+    email: user.email,
+    hashedPassword: hash(user.password),
+  };
+  let newUser;
+  const inviteCode = user.inviteCodeId
+    ? await getInviteById(user.inviteCodeId)
+    : null;
+  if (inviteCode) {
+    // inviteCodeId exists and valid invite code fetched
+    newUser = await prisma.user.update({
+      data: loginInfo,
+      where: {
+        id: inviteCode.user_id,
+      },
+    });
+  } else {
+    // signing up without an invite code
+    newUser = await prisma.user.create({
+      data: loginInfo,
+    });
+  }
+  if (!newUser) {
+    return null;
+  }
+  return sanitizeUser(newUser);
+};
+
+const handler = async (
+  req: NextApiRequest,
+  res: NextApiResponse
+): Promise<void> => {
   try {
-    if (!Array.isArray(sampleUserData)) {
-      throw new Error("Cannot find user data");
+    const expectedBody = Joi.object({
+      name: Joi.string().required(),
+      email: Joi.string().required(),
```

Comment:
We can chain the [email](https://joi.dev/api/?v=17.2.1#stringemailoptions) validator too!

```suggestion
      email: Joi.string().email().required(),
```